# Variables, Lists & Dictionaries: the three containers everything else is built on

**▶ 01 Core Python** · 02 Pandas · 03 Cleaning · 04 Transformation · 05 Feature Engineering · 06 Regression · 07 Model Prep · 08 Case Studies · 09 Deployment

`01_Python_Fundamentals/01_variables_lists_and_dictionaries.ipynb`

---

### In one paragraph (no jargon)

Before you can analyse data you need somewhere to *put* it. Python gives you three everyday containers: a **variable** (one labelled box), a **list** (an ordered row of boxes, like a column in Excel) and a **dictionary** (a labelled lookup table, like a row of a spreadsheet where each value has a column name). Almost every pandas operation you'll meet later is one of these three, scaled up. Work through the 30 short business questions below and you'll have covered the entire vocabulary.

### After this notebook you can

- Do arithmetic and comparisons on business figures and print them in a readable sentence
- Create, read, update, delete and search a list: the four things you'll be asked to do
- Do the same for dictionaries, including looping over keys and values together
- Build a dictionary from two parallel lists in one line (a very common exam step)


### What's inside

1. Part 1: Numbers, operators and printing (Q1-Q10)
2. ⚡ Building a formatted report
3. Part 2: Lists, the workhorse (Q11-Q20)
4. Part 3: Dictionaries, labelled data (Q21-Q30)
5. ⚡ zip + dict comprehension, Counter and defaultdict
6. Exam quick-reference

---

> **▶ Runs on its own.** The next cell is the only setup you need. It imports the
> libraries and loads the data. If the `datasets/` folder isn't where it expects,
> it rebuilds an equivalent dataset in memory, so **every cell below still runs**,
> which is handy if you copy this single `.ipynb` somewhere else.
>
> **▶ Reading the cells.** Code comments explain *what the line does*; the text
> blocks explain *why you'd do it*. Look for these markers:
> `# WHAT:` a plain-English translation · `# WHY:` the reason it matters ·
> `# 🔧 CHANGE THIS:` the knob to turn when the exam question differs ·
> **⚡ Beyond the syllabus** = optional, higher-mark techniques.

### Jargon buster

| Word you'll see | What it actually means |
|---|---|
| **variable** | A labelled box that holds a value. `price = 250` puts 250 in a box called `price`. |
| **list** `[ ]` | An ordered shopping list of values. You can add, remove and reorder items. |
| **tuple** `( )` | Like a list, but frozen: once made it cannot be changed. Good for fixed records. |
| **dictionary** `{key: value}` | A labelled lookup table, like a contacts app: look up "Ravi", get his number. |
| **index** | The position number of an item. Python counts from **0**, so the 1st item is at index 0. |
| **function** | A named recipe you can re-run with different ingredients. |
| **argument / parameter** | The ingredients you hand to a function. |
| **return** | The answer a function hands back to you. |
| **iterate / loop** | Do the same thing once for every item in a list. |
| **boolean** | A yes/no value: `True` or `False`. |
| **string** `'text'` | Text. Always wrapped in quotes. |
| **f-string** `f"{x}"` | A sentence with live values slotted in: the report-writing tool. |

In [1]:
# =============================================================================
# SETUP — run this cell first. Everything below depends only on this.
# =============================================================================
# WHAT: these are Python's own built-in toolboxes; nothing needs installing.
import math                     # square roots, rounding, constants
import statistics               # mean / median without pandas
from collections import Counter, defaultdict   # counting and grouping helpers
from functools import reduce    # folds a whole list down to one value

# -----------------------------------------------------------------------------
# Exam-safe replacement for input()
# -----------------------------------------------------------------------------
# WHY: a notebook that calls input() STOPS and waits for typing. If the examiner
#      clicks "Run All", it hangs. `ask()` behaves exactly like input() but plays
#      back a scripted answer instead, so the notebook always runs end-to-end.
# 🔧 CHANGE THIS: set INTERACTIVE = True if you WANT to type answers yourself.
INTERACTIVE = False
_scripted_answers = iter(['42', '7', '3', '15', '100', '5', '2', '9', '1', '0'])

def ask(prompt=''):
    """Stand-in for input() that never blocks a Run All."""
    if INTERACTIVE:
        return input(prompt)
    value = next(_scripted_answers, '0')
    print(f"{prompt}{value}    <- simulated keyboard input")
    return value

print("Setup complete — you can now run any cell below in any order.")

Setup complete — you can now run any cell below in any order.


## Part 1: Numbers, operators and printing

A *variable* is just a name you stick on a value so you can reuse it. The operators are the ones from school maths, plus two worth memorising: `%` gives the **remainder** after division (useful for 'does this divide evenly?' questions) and `**` means **to the power of** (useful for growth forecasts).

Notice every answer below ends with a `print(f"...")`. That `f` lets you drop a live value straight into a sentence: in an exam, a labelled answer reads far better than a bare number.

**Q1. How can you calculate the total cost if a product price is 250 and shipping charge is 40??**

In [2]:
# Solution for Q1
product_price = 250  # base selling price of the product
shipping_charge = 40  # flat logistics fee
total_cost = product_price + shipping_charge  # combine fixed and variable components
print(f"Total payable amount: ₹{total_cost}")

Total payable amount: ₹290


**Q2. How can you find the remainder when total sales 1234 are divided equally among 12 salespeople??**

In [3]:
# Solution for Q2
total_sales = 1234  # overall revenue collected
salespeople = 12  # number of staff to split among
remainder = total_sales % salespeople  # modulus captures the leftover amount
print(f"Remainder after equal split: {remainder}")

Remainder after equal split: 10


**Q3. What is the Python expression to calculate 7 raised to the power of 4 for forecasting growth??**

In [4]:
# Solution for Q3
base = 7
exponent = 4
forecasted_growth = base ** exponent  # exponentiation operator models repeated multiplication
print(f"7 raised to the power of 4 = {forecasted_growth}")

7 raised to the power of 4 = 2401


In [5]:
# --- Another valid way to write the same answer -------------------
# WHY show two? Examiners give credit for either. Pick whichever you
# can reproduce from memory under time pressure.
7 ** 4

2401

**Q4. How can you check if two budgets `budget1=50000` and `budget2=75000` are equal??**

In [6]:
# Solution for Q4
budget1 = 50000
budget2 = 75000
are_equal = budget1 == budget2  # equality comparison returns a boolean
print(f"Budgets equal? {are_equal}")

Budgets equal? False


**Q5. How can you find the maximum revenue between two branches with revenues 85000 and 92000??**

In [7]:
# Solution for Q5
branch_a = 85000
branch_b = 92000
max_revenue = max(branch_a, branch_b)  # `max` selects the higher value
print(f"Highest branch revenue: ₹{max_revenue}")

Highest branch revenue: ₹92000


**Q6. How can you check if a stock quantity 360 is divisible by carton size 12??**

In [8]:
# Solution for Q6
stock_quantity = 360
carton_size = 12
is_divisible = stock_quantity % carton_size == 0  # divisible when no remainder exists
print(f"Can stock be evenly boxed? {is_divisible}")

Can stock be evenly boxed? True


**Q7. How can you round the monthly expense 4578.678 to the nearest integer??**

In [9]:
# Solution for Q7
monthly_expense = 4578.678
rounded_expense = round(monthly_expense)  # nearest integer for budgeting overview
print(f"Rounded monthly expense: ₹{rounded_expense}")

Rounded monthly expense: ₹4579


**Q8. How can you compute the average customer rating from `[4.2, 3.8, 4.5, 4.0]`??**

In [10]:
# Solution for Q8
ratings = [4.2, 3.8, 4.5, 4.0]
average_rating = sum(ratings) / len(ratings)  # arithmetic mean
print(f"Average customer rating: {average_rating:.2f}")

Average customer rating: 4.12


**Q9. How can you swap values of two department budgets `deptA=120000` and `deptB=80000`??**

In [11]:
# Solution for Q9
dept_a = 120000
dept_b = 80000
dept_a, dept_b = dept_b, dept_a  # tuple unpacking swaps values in one step
print(f"Department A budget: ₹{dept_a}")
print(f"Department B budget: ₹{dept_b}")

Department A budget: ₹80000
Department B budget: ₹120000


**Q10. How can you check the data type of `invoice_id='INV1023'`??**

In [12]:
# Solution for Q10
invoice_id = 'INV1023'
data_type = type(invoice_id).__name__  # class name gives human-readable type
print(f"`invoice_id` stores a {data_type}")

`invoice_id` stores a str


### ⚡ Beyond the syllabus: build a formatted report, not a wall of prints

Examiners reward output that a manager could actually read. Three tools do the heavy lifting: **format specifiers** inside f-strings (`:,.2f` = thousands separator + 2 decimals), **`:>10`** to right-align a column, and a loop to print rows. Same numbers, far better presentation.

In [13]:
# WHAT: a tiny report generator. Same data, presentation-ready output.
branch_revenue = {'Delhi': 8500000, 'Mumbai': 12400000, 'Bengaluru': 9730500}
total = sum(branch_revenue.values())

print(f"{'BRANCH':<12}{'REVENUE (₹)':>16}{'SHARE':>9}")   # <12 = left-align in 12 chars, >16 = right-align
print("-" * 37)
for branch, revenue in branch_revenue.items():
    share = revenue / total * 100
    #  :>16,.0f  ->  right-aligned, thousands separators, no decimals
    print(f"{branch:<12}{revenue:>16,.0f}{share:>8.1f}%")
print("-" * 37)
print(f"{'TOTAL':<12}{total:>16,.0f}{100:>8.1f}%")

BRANCH           REVENUE (₹)    SHARE
-------------------------------------
Delhi              8,500,000    27.8%
Mumbai            12,400,000    40.5%
Bengaluru          9,730,500    31.8%
-------------------------------------
TOTAL             30,630,500   100.0%


## Part 2: Lists, the workhorse

A list is an ordered sequence written in square brackets. Two rules trip everyone up:

- **Counting starts at 0.** `clients[0]` is the *first* client, `clients[2]` is the *third*.
- **Slices stop one short.** `inventory[2:5]` gives positions 2, 3 and 4, never 5. Read the colon as "up to, but not including".

Negative numbers count backwards from the end, so `orders[-1]` is always the last item however long the list is.

**Q11. Given the list of clients `clients=['Tata','Infosys','Reliance','Adani']`, how can you access the second element??**

In [14]:
# Solution for Q11
clients = ['Tata', 'Infosys', 'Reliance', 'Adani']
second_client = clients[1]  # zero-based indexing
print(f"Second client in the list: {second_client}")

Second client in the list: Infosys


**Q12. How can you get the last order from the list `orders=['ORD1','ORD2','ORD3','ORD4']` using negative indexing??**

In [15]:
# Solution for Q12
orders = ['ORD1', 'ORD2', 'ORD3', 'ORD4']
last_order = orders[-1]  # negative index counts from the end
print(f"Most recent order: {last_order}")

Most recent order: ORD4


**Q13. How can you retrieve items from index 2 to 4 in the list `inventory=[100,200,300,400,500]`??**

In [16]:
# Solution for Q13
inventory = [100, 200, 300, 400, 500]
subset = inventory[2:5]  # slice from index 2 up to but not including 5
print(f"Items at positions 2-4: {subset}")

Items at positions 2-4: [300, 400, 500]


**Q14. How can you update the third element in `departments=['HR','Finance','Sales']` to 'Marketing'??**

In [17]:
# Solution for Q14
departments = ['HR', 'Finance', 'Sales']
departments[2] = 'Marketing'  # mutate the specific index
print(f"Updated departments: {departments}")

Updated departments: ['HR', 'Finance', 'Marketing']


**Q15. How can you add a new product 'Laptop' to the list `products=['Mouse','Keyboard']`??**

In [18]:
# Solution for Q15
products = ['Mouse', 'Keyboard']
products.append('Laptop')  # append adds an item at the end
print(f"Expanded catalogue: {products}")

Expanded catalogue: ['Mouse', 'Keyboard', 'Laptop']


**Q16. How can you remove 'Keyboard' from `products=['Mouse','Keyboard','Monitor']`??**

In [19]:
# Solution for Q16
products = ['Mouse', 'Keyboard', 'Monitor']
products.remove('Keyboard')  # remove deletes the first matching entry
print(f"Inventory after removal: {products}")

Inventory after removal: ['Mouse', 'Monitor']


**Q17. How can you calculate the number of employees in the list `employees=['John','Priya','Ali','Meera']`??**

In [20]:
# Solution for Q17
employees = ['John', 'Priya', 'Ali', 'Meera']
headcount = len(employees)  # length corresponds to number of items
print(f"Employee count: {headcount}")

Employee count: 4


**Q18. How can you reverse the order of `transactions=[1000,2000,3000,4000]`??**

In [21]:
# Solution for Q18
transactions = [1000, 2000, 3000, 4000]
transactions.reverse()  # in-place reversal for chronological review
print(f"Transactions in reverse order: {transactions}")

Transactions in reverse order: [4000, 3000, 2000, 1000]


**Q19. How can you check if 'Infosys' exists in the list `clients=['Tata','Infosys','Reliance']`??**

In [22]:
# Solution for Q19
clients = ['Tata', 'Infosys', 'Reliance']
exists = 'Infosys' in clients  # membership test
print(f"Is Infosys a client? {exists}")

Is Infosys a client? True


**Q20. How can you generate a list of invoice numbers from 101 to 110 using `range`??**

In [23]:
# Solution for Q20
invoice_numbers = list(range(101, 111))  # range upper bound is exclusive
print(f"Generated invoice IDs: {invoice_numbers}")

Generated invoice IDs: [101, 102, 103, 104, 105, 106, 107, 108, 109, 110]


## Part 3: Dictionaries, labelled data

A dictionary stores **key → value** pairs. Instead of remembering that position 3 holds the location, you just ask for `supplier['location']`. This is exactly how a pandas row works, so getting comfortable here pays off twice.

The safe way to read a key is `supplier.get('rating', 'not rated')`: if the key is missing you get your fallback instead of a crash. Under exam pressure that one habit prevents a lot of red error text.

**Q21. How can you create a dictionary for a supplier with keys 'name','location' and values 'ABC Ltd','Mumbai'??**

In [24]:
# Solution for Q21
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
print(f"Supplier record: {supplier}")

Supplier record: {'name': 'ABC Ltd', 'location': 'Mumbai'}


**Q22. How can you access the 'location' value from dictionary `supplier={'name':'ABC Ltd','location':'Mumbai'}`??**

In [25]:
# Solution for Q22
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
location = supplier['location']  # direct key lookup
print(f"Supplier location: {location}")

Supplier location: Mumbai


**Q23. How can you add a new key 'rating' with value 4.5 in dictionary `supplier={'name':'ABC Ltd','location':'Mumbai'}`??**

In [26]:
# Solution for Q23
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
supplier['rating'] = 4.5  # assign a new key-value pair
print(f"Supplier with rating: {supplier}")

Supplier with rating: {'name': 'ABC Ltd', 'location': 'Mumbai', 'rating': 4.5}


**Q24. How can you update the value of 'location' to 'Delhi' in dictionary `supplier={'name':'ABC Ltd','location':'Mumbai'}`??**

In [27]:
# Solution for Q24
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
supplier['location'] = 'Delhi'  # overwrite existing key value
print(f"Updated supplier: {supplier}")

Updated supplier: {'name': 'ABC Ltd', 'location': 'Delhi'}


**Q25. How can you remove the key 'location' from dictionary `supplier={'name':'ABC Ltd','location':'Mumbai'}`??**

In [28]:
# Solution for Q25
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
removed_location = supplier.pop('location')  # pop removes and returns the value
print(f"Removed location: {removed_location}")
print(f"Supplier after removal: {supplier}")

Removed location: Mumbai
Supplier after removal: {'name': 'ABC Ltd'}


**Q26. How can you get all the keys from dictionary `supplier={'name':'ABC Ltd','location':'Mumbai'}`??**

In [29]:
# Solution for Q26
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
keys = list(supplier.keys())  # convert to list for display
print(f"Dictionary keys: {keys}")

Dictionary keys: ['name', 'location']


**Q27. How can you get all the values from dictionary `supplier={'name':'ABC Ltd','location':'Mumbai'}`??**

In [30]:
# Solution for Q27
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
values = list(supplier.values())  # values view
print(f"Dictionary values: {values}")

Dictionary values: ['ABC Ltd', 'Mumbai']


**Q28. How can you check if key 'rating' exists in dictionary `supplier={'name':'ABC Ltd','location':'Mumbai'}`??**

In [31]:
# Solution for Q28
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
has_rating = 'rating' in supplier  # membership on keys
print(f"Rating key present? {has_rating}")

Rating key present? False

**Q29. How can you loop through both keys and values of dictionary `supplier={'name':'ABC Ltd','location':'Mumbai'}`??**

In [32]:
# Solution for Q29
supplier = {'name': 'ABC Ltd', 'location': 'Mumbai'}
for key, value in supplier.items():
    print(f"{key.title()}: {value}")  # iterate pairs with formatted output

Name: ABC Ltd
Location: Mumbai


**Q30. How can you create a dictionary mapping product codes `['P01','P02']` to stock counts `[150,300]`??**

In [33]:
# Solution for Q30
product_codes = ['P01', 'P02']
stock_counts = [150, 300]
product_stock_map = dict(zip(product_codes, stock_counts))  # pair each code with its stock
print(f"Product stock mapping: {product_stock_map}")

Product stock mapping: {'P01': 150, 'P02': 300}


### ⚡ Beyond the syllabus: build a dictionary from two lists with `zip` and a dict comprehension

Q30 asks you to pair product codes with stock counts. The manual way is a loop; the professional way is one line. `zip` walks two lists side by side like a zipper, and a **dict comprehension** `{k: v for k, v in ...}` builds the dictionary as it goes. This exact pattern reappears constantly in pandas (`dict(zip(df.code, df.stock))`).

In [34]:
codes  = ['P01', 'P02', 'P03', 'P04']
stock  = [150, 300, 0, 42]

# THE ONE-LINER: zip pairs them up, the {k: v for ...} builds the dictionary
stock_by_code = {code: qty for code, qty in zip(codes, stock)}
print("Built in one line:", stock_by_code)

# Same idea, filtered — only products that need reordering
reorder = {code: qty for code, qty in zip(codes, stock) if qty < 100}
print("Needs reordering:  ", reorder)

# Flip it around: value -> key (handy for lookups in the other direction)
code_by_stock = {qty: code for code, qty in stock_by_code.items()}
print("Reversed mapping:  ", code_by_stock)

Built in one line: {'P01': 150, 'P02': 300, 'P03': 0, 'P04': 42}
Needs reordering:   {'P03': 0, 'P04': 42}
Reversed mapping:   {150: 'P01', 300: 'P02', 0: 'P03', 42: 'P04'}


### ⚡ Beyond the syllabus: counting things properly with `Counter` and `defaultdict`

"How many orders per city?" is the single most common data question there is. Writing it with a plain dictionary takes five lines and an `if key not in dict` check. `Counter` does it in one and hands you `.most_common()` for free, this is the pure-Python ancestor of `df.groupby().size()`.

In [35]:
cities = ['Delhi', 'Mumbai', 'Delhi', 'Pune', 'Mumbai', 'Delhi', 'Chennai']

# THE LONG WAY (what most people write)
counts_manual = {}
for c in cities:
    counts_manual[c] = counts_manual.get(c, 0) + 1   # .get(c, 0) avoids a KeyError
print("Manual   :", counts_manual)

# THE SHORT WAY
print("Counter  :", dict(Counter(cities)))
print("Top 2    :", Counter(cities).most_common(2))   # sorted for you, highest first

# GROUPING (not just counting) — defaultdict never raises KeyError
orders = [('Delhi', 1200), ('Mumbai', 900), ('Delhi', 450)]
by_city = defaultdict(list)          # any new key starts life as an empty list
for city, amount in orders:
    by_city[city].append(amount)     # no "if city not in by_city" needed
print("Grouped  :", dict(by_city))
print("City totals:", {c: sum(v) for c, v in by_city.items()})

Manual   :

 {'Delhi': 3, 'Mumbai': 2, 'Pune': 1, 'Chennai': 1}
Counter  : {'Delhi': 3, 'Mumbai': 2, 'Pune': 1, 'Chennai': 1}
Top 2    : [('Delhi', 3), ('Mumbai', 2)]
Grouped  : {'Delhi': [1200, 450], 'Mumbai': [900]}
City totals: {'Delhi': 1650, 'Mumbai': 900}


---

## Exam quick-reference

| To do this | Write this |
|---|---|
| Add / multiply / power | `a + b` · `a * b` · `a ** b` |
| Remainder (divisibility test) | `360 % 12 == 0` |
| Round to nearest whole number | `round(4578.678)` |
| Average of a list | `sum(vals) / len(vals)` |
| Swap two variables | `a, b = b, a` |
| First / last item | `lst[0]` · `lst[-1]` |
| Items 3rd to 5th | `lst[2:5]` (stops before 5) |
| Add / remove from a list | `lst.append(x)` · `lst.remove(x)` |
| Reverse a list | `lst[::-1]` (copy) or `lst.reverse()` (in place) |
| Is it in the list? | `'Infosys' in clients` |
| Numbers 101…110 | `list(range(101, 111))` |
| Read a key safely | `d.get('rating', 'N/A')` |
| Add / update / delete a key | `d['k'] = v` · `del d['k']` |
| All keys / values / pairs | `d.keys()` · `d.values()` · `d.items()` |
| Loop keys **and** values | `for k, v in d.items():` |
| Two lists → dictionary | `dict(zip(keys, values))` |

### Adapting this in the exam

- Different numbers in the question? Only the values on the right of the `=` change. The structure never does.
- Asked for a *count* instead of a *total*? Swap `sum(...)` for `len(...)`.
- Asked for the smallest instead of the largest? Swap `max` for `min`. Everything else stays.

### Traps that cost marks

- `range(101, 110)` stops at **109**. To include 110 you need `range(101, 111)`.
- `lst.remove('x')` deletes the *first* match only, and crashes if `'x'` isn't there. Check with `if 'x' in lst:` first.
- `lst.reverse()` returns `None`; it changes the list in place. `print(lst.reverse())` prints `None`, which looks like a bug but isn't.
- `d['missing']` crashes with a `KeyError`. `d.get('missing')` quietly returns `None`.
- Dictionary keys must be unique. Assigning the same key twice silently overwrites it, with no warning.